sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -247,58 +259,137 @@ object CTESubstitution extends Rule[LogicalPlan] {
+        .map(_._2.copy(child = substituted, recursive = recursionFound))
+        .getOrElse(CTERelationDef(substituted))
       if (!alwaysInline) {
         cteDefs += cteRelation
       }
+
+      // From this point any reference to the definition is non-recursive.
+      val nonRecursiveCTERelation = if (cteRelation.recursive) {
+        cteRelation.copy(recursive = false)
+      } else {
+        cteRelation
+      }
+
       // Prepending new CTEs makes sure that those have higher priority over outer ones.
-      resolvedCTERelations +:= (name -> cteRelation)
+      resolvedCTERelations +:= (name -> nonRecursiveCTERelation)
     }
     resolvedCTERelations
   }
 
+  /**
+   * @return tuple of the logical plan and recursionFound boolean.
+   */
   private def resolveWithCTERelations(
       table: String,
       alwaysInline: Boolean,
       cteRelations: Seq[(String, CTERelationDef)],
-      unresolvedRelation: UnresolvedRelation): LogicalPlan = {
-    cteRelations
+      unresolvedRelation: UnresolvedRelation): (LogicalPlan, Boolean) = {
+    var recursionFound = false
+    val cteRelationsOut = cteRelations
       .find(r => conf.resolver(r._1, table))
       .map {
         case (_, d) =>
           if (alwaysInline) {
             d.child
           } else {
+            if (d.recursive) {
+              recursionFound = true
+            }
             // Add a `SubqueryAlias` for hint-resolving rules to match relation names.
-            SubqueryAlias(table, CTERelationRef(d.id, d.resolved, d.output, d.isStreaming))
+            SubqueryAlias(table,
+              CTERelationRef(d.id, d.resolved, d.output, d.isStreaming, recursive = d.recursive))
           }
       }
       .getOrElse(unresolvedRelation)
+    (cteRelationsOut, recursionFound)
   }
 
+  /**
+   * @return tuple of the logical plan and recursionFound boolean.
+   */
   private def substituteCTE(
       plan: LogicalPlan,
       alwaysInline: Boolean,
-      cteRelations: Seq[(String, CTERelationDef)]): LogicalPlan = {
-    plan.resolveOperatorsUpWithPruning(
+      cteRelations: Seq[(String, CTERelationDef)],
+      recursiveCTERelation: Option[(String, CTERelationDef)]): (LogicalPlan, Boolean) = {
+    var recursionFound = false
+    val substituted = plan.resolveOperatorsUpWithPruning(
         _.containsAnyPattern(RELATION_TIME_TRAVEL, UNRESOLVED_RELATION, PLAN_EXPRESSION,
           UNRESOLVED_IDENTIFIER)) {
       case RelationTimeTravel(UnresolvedRelation(Seq(table), _, _), _, _)
         if cteRelations.exists(r => plan.conf.resolver(r._1, table)) =>
         throw QueryCompilationErrors.timeTravelUnsupportedError(toSQLId(table))
 
       case u @ UnresolvedRelation(Seq(table), _, _) =>
-        resolveWithCTERelations(table, alwaysInline, cteRelations, u)
+        val resolved = resolveWithCTERelations(table, alwaysInline,
+          (recursiveCTERelation ++ cteRelations).toSeq, u)
+        recursionFound = resolved._2
```

Comment:
shouldn't it be `recursionFound = recursionFound || resolved._2`?

Reply:
This should be corrected with the new commit. Thanks for the suggestion.